ipamutils/utils.go

```diff
@@ -48,3 +66,60 @@ func initGranularPredefinedNetworks() []*net.IPNet {
 	}
 	return pl
 }
+
+// InitAddressPools allows to initialize the local and global scope predefined
+// address pools to the desired values. It fails is invalid input is passed
+// or if the predefined pools were already initialized.
+func InitAddressPools(list []*PredefinedPools) error {
+	localPools := make([]*net.IPNet, 0, len(list))
+
+	for _, p := range list {
+		if p == nil {
+			continue
+		}
+		_, b, err := net.ParseCIDR(p.Base)
+		if err != nil {
+			return fmt.Errorf("invalid base pool %q: %v", p.Base, err)
+		}
+		ones, _ := b.Mask.Size()
+		if p.Size <= 0 || p.Size < ones {
+			return fmt.Errorf("invalid pools size: %d", p.Size)
+		}
+		localPools = append(localPools, initPools(p.Size, b)...)
+	}
+	if localPools == nil {
+		localPools = initBroadPredefinedNetworks()
+	}
+	PredefinedBroadNetworks = localPools
+	return nil
+}
+
+func initPools(size int, base *net.IPNet) []*net.IPNet {
+	one, bits := base.Mask.Size()
+	mask := net.CIDRMask(size, bits)
+	n := 1 << uint(size-one)
+	s := uint(bits - size)
+	list := make([]*net.IPNet, 0, n)
+
+	for i := 0; i < n; i++ {
+		ip := copyIP(base.IP)
+		addIntToIP(ip, uint(i<<s))
```

Comment:
@ctelfer-docker good point. But since we dont have default/pre-defined subnets for ipv6 (pls cross check), this particular PR is exempt since it addresses the pre-defined subnet functionality. wdyt ?

Reply:
(pls consider this an answer to both of your queries @mavenugo)
Looking at ipam/allocator.go I see the following: 
```go
func (a *Allocator) getPredefinedPool(as string, ipV6 bool) (*net.IPNet, error) {
        var v ipVersion
        v = v4
        if ipV6 {
                v = v6
        }
... // snip code that ensures that we are only requesting from the pre-defined network key strings
        for _, nw := range a.getPredefineds(as) {
                if v != getAddressVersion(nw.IP) {
                        continue
                }
...
        }
```
As one would expect from the name, this function (via other tracing) does get called for a request for a pre-defined network.  So to me it appears that libnetwork's default IPAM code was written with the intention of having 

Now it's certainly true that libnetwork doesn't currently allocate any IPv6 pre-defined networks and so `docker network create` without `--subnet=` will never get a successful hit out of the above function and would thus always fail.  (unless there is some codepath that adds more networks to the predefined set that I am not yet aware of).  So from that perspective I agree that we aren't losing  any functionality with this PR as is.

On the other hand, this PR is specifically to allow the user to specify the pre-defined subnets.  If we, with a few lines of code change, can cleanly support v6 as well as v4, my opinion would be that we should.  (and the above code snippet seems to indicate that the allocator is already "primed" for the change.)  But I certainly defer to more experienced reviewers here.

Reply:
I'm wondering if a better solution, maybe as a follow up for this PR would be good to rework it eventually making it on demand based. Is there any real advantage in splitting all the possible networks on init?